libraries/lib-module-manager/PluginHost.cpp

```diff
@@ -0,0 +1,220 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file PluginHost.cpp
+
+  @author Vitaly Sverchinsky
+
+  Part of lib-module-manager library
+
+**********************************************************************/
+
+#include "PluginHost.h"
+
+#include <wx/app.h>
+#include <wx/log.h>
+#include <wx/module.h>
+#include <wx/process.h>
+
+#include "BasicUI.h"
+#include "PathList.h"
+#include "FileNames.h"
+#include "ModuleManager.h"
+#include "IPCClient.h"
+#include "PlatformCompatibility.h"
+#include "PluginManager.h"
+
+namespace
+{
+   //Attempts to instantiate plugin module and put plugin descriptors into result
+   void Discover(details::PluginValidationResult& result, const wxString& providerId, const wxString& pluginPath)
+   {
+      try
+      {
+         if(auto provider = ModuleManager::Get().CreateProviderInstance(providerId, wxEmptyString))
+         {
+            TranslatableString errorMessage{};
+
+            provider->DiscoverPluginsAtPath(pluginPath, errorMessage, [&result](PluginProvider *provider, ComponentInterface *ident)
+            {
+               //Workaround: use DefaultRegistrationCallback to create all descriptors for us
+               //and then put a copy into result
+               auto id = PluginManager::DefaultRegistrationCallback(provider, ident);
+               if(const auto desc = PluginManager::Get().GetPlugin(id))
+                  result.Add(PluginDescriptor { *desc });
+               return id;
+            });
+            if(!errorMessage.empty())
+               result.SetError(errorMessage.Translation());
+         }
+         else
+            result.SetError("provider not found");
+      }
+      catch(...)
+      {
+         result.SetError("unknown error");
+      }
+   }
+}
+
+PluginHost::PluginHost()
+{
+   FileNames::InitializePathList();
+
+   wxFileName configFileName{ FileNames::Configuration() };
+   auto pConfig = std::make_unique<FileConfig>(
+      AppName, wxEmptyString, configFileName.GetFullPath(),
+      wxEmptyString, wxCONFIG_USE_LOCAL_FILE);
+   pConfig->Init();
+   InitPreferences(std::move(pConfig));
+
+   auto& moduleManager = ModuleManager::Get();
+   moduleManager.Initialize();
+   moduleManager.DiscoverProviders();
+
+   mClient = std::make_unique<IPCClient>(*this);
+}
+
+void PluginHost::OnConnect(IPCChannel& channel) noexcept
+{
+   std::lock_guard lck(mSync);
+   mChannel = &channel;
+}
+
+void PluginHost::OnDisconnect() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnConnectionError() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnDataAvailable(const void* data, size_t size) noexcept
+{
+   try
+   {
+      mInputMessageReader.ConsumeBytes(data, size);
+      if(mInputMessageReader.CanPop())
+      {
+         {
+            std::lock_guard lck(mSync);
+            assert(!mRequest);
```

Comment:
An assertion without proof.  To fix it, improve documentation to explain that a request must be sent only after responses of all previous requests have been received by the client -- otherwise, requests may be lost:  they don't queue up.

That makes this class more understandable without reading the code that uses the class -- and that code doesn't exist yet, at the commit introducing this.